### DISCUSSION TITLE
How do I build an abstract but efficient UI with Riverpod?

### DISCUSSION THREAD

**Discussion by Maldus512** [rrousselGit/riverpod#2905] [Q&A] [answered]
I need to develop two separate application that intersect in one functionality: being able to display and modify a certain data structure `A`. `A` has multiple fields that require specific widgets, but the UI that both applications should display is more or less the same. Also, both applications work with their own instance of `A`.

I would like to implement the UI for `A` as a separate library upon which both applications will depend and I'm struggling on how to properly structure it, especially in regard of the data providers.

At first I tried defining `StatelessWidgets` that display the fields of `A` by receiving them as parameters and modifying `A` via callbacks. This approach works, but after some experiments I realized that it results in several unnecessary UI rebuilds because the `watch` points are not inside the leaf widget but higher up:

```dart
class HomePage extends ConsumerWidget {
  const HomePage({super.key});

  @override
  Widget build(BuildContext context, WidgetRef ref) {
       return Column(children:[
            GenericSettingWidget(
              "First",
              ref.watch(aProvider.select((s) => s.firstSetting)),
              (value) =>
                  ref.read(aProvider.notifier).changeFirstSetting(value),
            ),
            GenericSettingWidget(
              "Second",
              ref.watch(aProvider.select((s) => s.secondSetting)),
              (value) => ref
                  .read(aProvider.notifier)
                  .changeSecondSetting(value),
            ),
         ],
      );
   }
}
```

Here, `GenericSettingWidget` is part of the library, while `HomePage` is a screen for one of the applications. `aProvider` lives in the application, outside of the library.

Since `aProvider` is `watch`ed through the `WidgetRef` of `HomePage`, the whole widget tree is rebuilt on every change.

In working like this I was trying to keep `A` instances and/or providers out of the library, but now I'm thinking that maybe that should not be the case.
Should I have a late `aProvider` inside the library that must be properly populated by the dependent application before using the  widgets?

If not, what is the typical solution to this problem using Riverpod? How can non-trivial widgets that display and modify generic data be implemented in a way that minimizes rebuilds?

**Comment by noriHanda** [comment]
@Maldus512 I'm not sure this fits in your exact case, but it seems like defining `aProvider` as a family provider should do. Something like this:

```dart
@riverpod
class A extends _$A {
  @override
  int build(int value) {
    return value;
  }
  <your methods here>
}
```

Passing a different parameter to the provider generated with the above code will generate a separate provider.

P.S.
Using syntax like this
\```dart
your code
\```
would be very helpful reading your code on GitHub.

**Comment by Maldus512** [comment]
@noriHanda I'm not sure I follow. On one hand I'm not familiar with the `@riverpod` annotation (or code generation in general) and I couldn't find references about building providers out of classes in the documentation (I'm probably missing an obvious part, sorry); on the other I don't think family providers help me here.
Mi issue is not about having multiple instances of one provider, I'm trying to understand where it is best to define a provider when building a UI library.

**Comment by snapsl** [comment, marked as answer]
@Maldus512 So currently the complete `HomePage` widget rebuilds every time one setting is updated since you define `ref.watch` here. 
A simple solution is to wrap each each entry in a consumer widget.

```dart
Consumer(
      builder: (context, ref, child) {
        final firstSetting = ref.watch(aProvider.select((s) => s.firstSetting));

        return GenericSettingWidget(
          'First',
          firstSetting,
          (value) => ref.read(aProvider.notifier).changeFirstSetting(value),
        );
      },
);
```
Or pass the `ProviderListenable` of `aProvider.select((s) => s.firstSetting)` to the `GenericSettingWidget` in order to use `ref.watch` there.

**Reply by Maldus512** [reply]
This works, thank you!
I think I will pass some `ProviderListenable`s as parameters to the widget; wrapping my widgets around a Consumer works for small UI pieces, but if `GenericSettingWidget` ended up displaying more than one property I could benefit from more selective rebuilds.
